File: animations/causal_anim/timeline.py
"""
Dual-clock timeline: causal ticks (τ) and presentation seconds (t).

Python wrapper that mirrors the Rust `Timeline` and also drives the
frame-level interpolation for the Python scene loop.
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import List


@dataclass
class TimeSegment:
    kind: str          # "rush" | "slow_motion" | "pause" | "normal"
    ticks: int = 0
    duration: float = 0.0
    ticks_per_second: float = 10.0
    causal_start: int = 0
    presentation_start: float = 0.0


class Timeline:
    """Manages the mapping τ ↔ t."""

    def __init__(self) -> None:
        self._segments: List[TimeSegment] = []
        self._current_causal: int = 0
        self._current_presentation: float = 0.0
        self._default_pace: float = 10.0

    # ── Segment builders ──────────────────────────────────────────────

    def rush(self, ticks: int, duration: float) -> None:
        """Compress *ticks* causal steps into *duration* seconds."""
        self._segments.append(TimeSegment(
            kind="rush", ticks=ticks, duration=duration,
            causal_start=self._current_causal,
            presentation_start=self._current_presentation,
        ))
        self._current_causal += ticks
        self._current_presentation += duration

    def slow_motion(self, ticks: int, duration: float) -> None:
        """Stretch *ticks* causal steps over *duration* seconds."""
        self._segments.append(TimeSegment(
            kind="slow_motion", ticks=ticks, duration=duration,
            causal_start=self._current_causal,
            presentation_start=self._current_presentation,
        ))
        self._current_causal += ticks
        self._current_presentation += duration

    def pause(self, duration: float) -> None:
        """Freeze the graph; advance only presentation time."""
        self._segments.append(TimeSegment(
            kind="pause", duration=duration,
            causal_start=self._current_causal,
            presentation_start=self._current_presentation,
        ))
        self._current_presentation += duration

    def set_pace(self, ticks_per_second: float) -> None:
        """Set the default pace for subsequent ``wait_ticks`` calls."""
        self._default_pace = ticks_per_second

    def wait_ticks(self, ticks: int) -> None:
        """Advance *ticks* causal steps at the current default pace."""
        duration = ticks / self._default_pace
        self._segments.append(TimeSegment(
            kind="normal", ticks=ticks, duration=duration,
            ticks_per_second=self._default_pace,
            causal_start=self._current_causal,
            presentation_start=self._current_presentation,
        ))
        self._current_causal += ticks
        self._current_presentation += duration

    # ── Queries ───────────────────────────────────────────────────────

    @property
    def total_duration(self) -> float:
        return self._current_presentation

    @property
    def total_causal_ticks(self) -> int:
        return self._current_causal

    def presentation_to_causal(self, t: float) -> float:
        """Map presentation time *t* to an interpolated causal tick."""
        for seg in reversed(self._segments):
            if t >= seg.presentation_start:
                dt = t - seg.presentation_start
                if seg.kind in ("rush", "slow_motion"):
                    progress = max(0.0, min(1.0, dt / seg.duration)) if seg.duration > 0 else 1.0
                    return seg.causal_start + progress * seg.ticks
                elif seg.kind == "pause":
                    return float(seg.causal_start)
                else:  # normal
                    return seg.causal_start + dt * seg.ticks_per_second
        return 0.0

    def segment_progress(self, t: float) -> float:
        """Fractional progress (0–1) within the active segment."""
        for seg in reversed(self._segments):
            if t >= seg.presentation_start:
                dt = t - seg.presentation_start
                return max(0.0, min(1.0, dt / seg.duration)) if seg.duration > 0 else 1.0
        return 0.0

File: animations/causal_anim_core/src/timeline.rs
//! Dual-clock timeline: causal ticks (τ) and presentation seconds (t).

/// The kind of pacing for a timeline segment.
#[derive(Debug, Clone)]
pub enum PaceKind {
    /// Compress many ticks into a short duration (fast-forward).
    Rush { ticks: u64, duration_secs: f64 },
    /// Stretch a few ticks over a long duration (slow motion).
    SlowMotion { ticks: u64, duration_secs: f64 },
    /// Freeze τ, advance only presentation time (for narration).
    Pause { duration_secs: f64 },
    /// Constant pace: N causal ticks per second of video.
    Normal { ticks_per_second: f64 },
}

/// A segment of the timeline with its pacing and starting clocks.
#[derive(Debug, Clone)]
pub struct TimeSegment {
    pub kind: PaceKind,
    pub causal_start: u64,
    pub presentation_start: f64,
}

impl TimeSegment {
    /// Presentation duration of this segment.
    pub fn duration_secs(&self) -> f64 {
        match &self.kind {
            PaceKind::Rush { duration_secs, .. }
            | PaceKind::SlowMotion { duration_secs, .. }
            | PaceKind::Pause { duration_secs } => *duration_secs,
            PaceKind::Normal { .. } => 0.0, // open-ended; closed by next segment
        }
    }

    /// Number of causal ticks in this segment.
    pub fn causal_ticks(&self) -> u64 {
        match &self.kind {
            PaceKind::Rush { ticks, .. } | PaceKind::SlowMotion { ticks, .. } => *ticks,
            PaceKind::Pause { .. } => 0,
            PaceKind::Normal { .. } => 0,
        }
    }
}

/// The dual-clock timeline.
///
/// Maintains an ordered list of segments and tracks running totals for
/// both clocks.
#[derive(Debug, Clone)]
pub struct Timeline {
    segments: Vec<TimeSegment>,
    current_causal: u64,
    current_presentation: f64,
    default_pace: f64, // ticks per second (for wait_ticks)
}

impl Timeline {
    pub fn new() -> Self {
        Self {
            segments: Vec::new(),
            current_causal: 0,
            current_presentation: 0.0,
            default_pace: 10.0,
        }
    }

    pub fn rush(&mut self, ticks: u64, duration_secs: f64) {
        self.segments.push(TimeSegment {
            kind: PaceKind::Rush { ticks, duration_secs },
            causal_start: self.current_causal,
            presentation_start: self.current_presentation,
        });
        self.current_causal += ticks;
        self.current_presentation += duration_secs;
    }

    pub fn slow_motion(&mut self, ticks: u64, duration_secs: f64) {
        self.segments.push(TimeSegment {
            kind: PaceKind::SlowMotion { ticks, duration_secs },
            causal_start: self.current_causal,
            presentation_start: self.current_presentation,
        });
        self.current_causal += ticks;
        self.current_presentation += duration_secs;
    }

    pub fn pause(&mut self, duration_secs: f64) {
        self.segments.push(TimeSegment {
            kind: PaceKind::Pause { duration_secs },
            causal_start: self.current_causal,
            presentation_start: self.current_presentation,
        });
        self.current_presentation += duration_secs;
    }

    pub fn set_pace(&mut self, ticks_per_second: f64) {
        self.default_pace = ticks_per_second;
    }

    /// Add a normal-pace segment that advances `ticks` causal steps at the
    /// currently configured pace.
    pub fn wait_ticks(&mut self, ticks: u64) {
        let duration = ticks as f64 / self.default_pace;
        self.segments.push(TimeSegment {
            kind: PaceKind::Normal { ticks_per_second: self.default_pace },
            causal_start: self.current_causal,
            presentation_start: self.current_presentation,
        });
        self.current_causal += ticks;
        self.current_presentation += duration;
    }

    pub fn total_duration(&self) -> f64 {
        self.current_presentation
    }

    pub fn total_causal_ticks(&self) -> u64 {
        self.current_causal
    }

    /// Map presentation time t → interpolated causal tick.
    pub fn presentation_to_causal(&self, t: f64) -> f64 {
        for seg in self.segments.iter().rev() {
            if t >= seg.presentation_start {
                let dt = t - seg.presentation_start;
                return match &seg.kind {
                    PaceKind::Rush { ticks, duration_secs }
                    | PaceKind::SlowMotion { ticks, duration_secs } => {
                        let progress = (dt / duration_secs).clamp(0.0, 1.0);
                        seg.causal_start as f64 + progress * *ticks as f64
                    }
                    PaceKind::Pause { .. } => seg.causal_start as f64,
                    PaceKind::Normal { ticks_per_second } => {
                        seg.causal_start as f64 + dt * ticks_per_second
                    }
                };
            }
        }
        0.0
    }

    /// Map presentation time t → fractional progress within the active
    /// segment (0.0 – 1.0).  Useful for easing animations.
    pub fn segment_progress(&self, t: f64) -> f64 {
        for seg in self.segments.iter().rev() {
            if t >= seg.presentation_start {
                let dt = t - seg.presentation_start;
                let dur = seg.duration_secs();
                return if dur > 0.0 { (dt / dur).clamp(0.0, 1.0) } else { 1.0 };
            }
        }
        0.0
    }

    pub fn segments(&self) -> &[TimeSegment] {
        &self.segments
    }
}

impl Default for Timeline {
    fn default() -> Self {
        Self::new()
    }
}
